File: frontend/src/api/ruonia.ts
import { apiClient } from './client';

export interface RuoniaRefreshResponse {
  status: 'ok' | 'error';
  message?: string;
  from_date?: string;
  to_date?: string;
  new_entries?: number;
  updated_entries?: number;
  total_entries?: number;
  updated?: boolean;
  error?: string;
}

export interface RuoniaRecord {
  'Дата ставки': string;
  'Ставка RUONIA, % годовых': number | null;
  'Объем сделок RUONIA, млрд руб.': number | null;
  'Количество сделок, ед.': number | null;
  'Минимальная процентная ставка, % годовых': number | null;
  '25-й процентиль по процентным ставкам, % годовых': number | null;
  '75-й процентиль по процентным ставкам, % годовых': number | null;
  'Максимальная процентная ставка, % годовых': number | null;
}

export interface RuoniaDataResponse {
  data: RuoniaRecord[];
  count: number;
  date_from: string | null;
  date_to: string | null;
}

/**
 * Fetch RUONIA data filtered by date range
 */
export const fetchRuoniaData = async (
  dateFrom?: string | null,
  dateTo?: string | null
): Promise<RuoniaDataResponse> => {
  const params: Record<string, string> = {};
  if (dateFrom) params.date_from = dateFrom;
  if (dateTo) params.date_to = dateTo;

  const response = await apiClient.get<RuoniaDataResponse>('/ruonia/data', { params });
  return response.data;
};

/**
 * Download RUONIA data as Markdown
 */
export const downloadRuoniaMarkdown = async (
  dateFrom?: string | null,
  dateTo?: string | null
): Promise<void> => {
  const params: Record<string, string> = {};
  if (dateFrom) params.date_from = dateFrom;
  if (dateTo) params.date_to = dateTo;

  const response = await apiClient.get('/ruonia/download/markdown', {
    params,
    responseType: 'blob',
  });

  // Create download link
  const blob = new Blob([response.data], { type: 'text/markdown;charset=utf-8;' });
  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;

  // Generate filename
  let filename = 'ruonia';
  if (dateFrom || dateTo) {
    if (dateFrom) filename += `_${dateFrom.replace(/\./g, '-')}`;
    if (dateTo) filename += `_${dateTo.replace(/\./g, '-')}`;
  } else {
    filename += '_all';
  }
  filename += '.md';

  link.download = filename;
  link.click();
  URL.revokeObjectURL(url);
};

/**
 * Request a refresh of RUONIA rate data from CBR
 */
export const refreshRuoniaData = async (): Promise<RuoniaRefreshResponse> => {
  const response = await apiClient.post<RuoniaRefreshResponse>('/ruonia/refresh');
  return response.data;
};

File: backend/app/routers/ruonia.py
"""Роутеры для работы с данными индикатора RUONIA.

Обрабатывают HTTP запросы: получение данных с фильтрами по датам (from, till),
экспорт в Markdown и обновление данных из ЦБ РФ. Параметры передаются в сервис
без обработки; вся бизнес-логика и преобразование данных — в service layer.
"""

import asyncio
from typing import Any, Optional

from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import Response

from app.models import RuoniaDataResponse
from app.services.ruonia_service import get_ruonia_service
from app.utils.logger import get_data_update_logger

router = APIRouter(prefix="/api/ruonia", tags=["ruonia"])


@router.get("/data", response_model=RuoniaDataResponse, response_model_by_alias=True)
async def get_ruonia_data(
    date_from: Optional[str] = Query(None, description="Start date DD.MM.YYYY"),
    date_to: Optional[str] = Query(None, description="End date DD.MM.YYYY"),
) -> RuoniaDataResponse:
    """Возвращает данные RUONIA с фильтрацией по диапазону дат.

    Параметры date_from и date_to передаются в сервис без изменения.
    Если не указаны — фильтр по соответствующей границе не применяется.
    """
    logger = get_data_update_logger()
    logger.info("[API /ruonia/data] Request received, date_from=%s, date_to=%s", date_from, date_to)
    try:
        ruonia_service = get_ruonia_service()
        result = await asyncio.to_thread(
            ruonia_service.get_ruonia_data,
            date_from=date_from,
            date_to=date_to,
        )
        logger.info("[API /ruonia/data] Success: %s records", result.count)
        return result
    except RuntimeError as exc:
        logger.error("[API /ruonia/data] RuntimeError: %s", exc)
        raise HTTPException(status_code=502, detail=str(exc)) from exc
    except Exception as exc:
        logger.error("[API /ruonia/data] ERROR: %s - %s", type(exc).__name__, exc)
        raise HTTPException(status_code=500, detail=f"Failed to get RUONIA data: {exc}") from exc


@router.get("/download/markdown")
async def download_ruonia_markdown(
    date_from: Optional[str] = Query(None, description="Start date DD.MM.YYYY"),
    date_to: Optional[str] = Query(None, description="End date DD.MM.YYYY"),
) -> Response:
    """Скачивает данные RUONIA в формате Markdown.

    Параметры date_from и date_to передаются в сервис без изменения.
    """
    logger = get_data_update_logger()
    logger.info("[API /ruonia/download/markdown] Request received, date_from=%s, date_to=%s", date_from, date_to)
    try:
        ruonia_service = get_ruonia_service()
        markdown_content = await asyncio.to_thread(
            ruonia_service.export_markdown,
            date_from=date_from,
            date_to=date_to,
        )
        filename = "ruonia"
        if date_from or date_to:
            if date_from:
                filename += f"_{date_from.replace('.', '-')}"
            if date_to:
                filename += f"_{date_to.replace('.', '-')}"
        else:
            filename += "_all"
        filename += ".md"
        return Response(
            content=markdown_content,
            media_type="text/markdown; charset=utf-8",
            headers={"Content-Disposition": f'attachment; filename="{filename}"'},
        )
    except Exception as exc:
        logger.error("[API /ruonia/download/markdown] ERROR: %s - %s", type(exc).__name__, exc)
        raise HTTPException(status_code=500, detail=f"Failed to generate Markdown: {exc}") from exc


@router.post("/refresh")
async def refresh_ruonia_data() -> dict[str, Any]:
    """Запускает обновление данных RUONIA из ЦБ РФ (инкрементальная загрузка)."""
    logger = get_data_update_logger()
    logger.info("[API /ruonia/refresh] Received request to refresh RUONIA data")
    try:
        ruonia_service = get_ruonia_service()
        result = await asyncio.to_thread(ruonia_service.update_ruonia_data)
        logger.info("[API /ruonia/refresh] Refresh completed: %s", result)
        return result
    except Exception as exc:
        logger.error("[API /ruonia/refresh] ERROR: %s - %s", type(exc).__name__, exc)
        raise HTTPException(status_code=500, detail=f"Failed to refresh RUONIA data: {exc}") from exc
